### PR TITLE
implement uploads for sponsor logos in admin

### PR DESCRIPTION
**Pull request by ewdurbin**

### CODE REVIEW

File: warehouse/admin/services.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        # This class should not be used in production, it's trivial for it to
+        # be used to read arbitrary files from the disk. It is intended ONLY
+        # for local development with trusted users. To make this clear, we'll
+        # raise a warning.
+        warnings.warn(
+            "LocalSponsorLogoStorage is intended only for use in development, you "
+            "should not use it in production due to the lack of safe guards "
+            "for safely locating files on disk.",
+            InsecureStorageWarning,
+        )
```
Comment (berinhard) [inline warehouse/admin/services.py:39]: Nice touch here!

File: warehouse/admin/views/sponsors.py [hunk 2]
```diff
@@ -95,6 +101,22 @@ def sponsor_list(request):
...
+            extension = os.path.splitext(request.POST["white_logo"].filename)[-1]
+            filename = f"{sponsor.id}-white-logo{extension}"
+            form.white_logo_url.data = storage.store(filename, fp.name)
+    if request.POST.get("color_logo", b"") != b"":
+        with tempfile.NamedTemporaryFile() as fp:
+            storage = request.find_service(ISponsorLogoStorage)
+            fp.write(request.POST["color_logo"].file.read())
+            extension = os.path.splitext(request.POST["white_logo"].filename)[-1]
+            filename = f"{sponsor.id}-color-logo{extension}"
+            form.color_logo_url.data = storage.store(filename, fp.name)
```
Comment (berinhard) [inline warehouse/admin/views/sponsors.py:118]: This is great! By updating `color/white_logo_url` we won't have to change anything were the logo is being displayed :+1:

File: warehouse/admin/views/sponsors.py [hunk 3]
```diff
@@ -95,6 +102,24 @@ def sponsor_list(request):
...
+            extension = os.path.splitext(request.POST["white_logo"].filename)[-1]
+            filename = f"{sponsor_name}-white-logo{extension}"
+            form.white_logo_url.data = storage.store(filename, fp.name)
+    if request.POST.get("color_logo") not in [None, b""]:
+        with tempfile.NamedTemporaryFile() as fp:
+            storage = request.find_service(ISponsorLogoStorage)
+            fp.write(request.POST["color_logo"].file.read())
+            extension = os.path.splitext(request.POST["color_logo"].filename)[-1]
+            filename = f"{sponsor_name}-color-logo{extension}"
+            form.color_logo_url.data = storage.store(filename, fp.name)
```
[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/admin/views/sponsors.py:120]: will address in #9794

File: warehouse/admin/views/sponsors.py [hunk 4]
```diff
@@ -151,15 +178,19 @@ def edit_sponsor(request):
...
-        sponsor = Sponsor(**form.data)
-        request.db.add(sponsor)
-        request.session.flash(
-            f"Added new sponsor '{sponsor.name}'",
-            queue="success",
-        )
-        redirect_url = request.route_url("admin.sponsor.list")
-        return HTTPSeeOther(location=redirect_url)
+    if request.method == "POST":
+        _handle_images(request, form)
```
[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/admin/views/sponsors.py:182]: form won't validate if the white_logo is being uploaded and we don't have a url yet :/

other option is to fingerprint each upload, which is fine, I think I'll will address in #9794 also.

[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/admin/views/sponsors.py:182]: (note fingerprinting will effectively mean that we'll "collect" a bunch of old logos in the bucket over time... which is probably OK?)

**Pull request merged**